snapcraft/tests/repo/test_snaps.py

```diff
@@ -0,0 +1,399 @@
+        query = parse.parse_qs(parsed_url.query)
+        status_code = 404
+        params = {}
+
+        if query['name'][0] == 'fake-snap':
+            status_code = 200
+            params = {'channels': {
+                'latest/stable': {'confinement': 'strict'},
+                'classic/stable': {'confinement': 'classic'},
+                'strict/stable': {'confinement': 'strict'},
+                'devmode/stable': {'confinement': 'devmode'},
+            }}
+        if query['name'][0] == 'new-fake-snap':
+            status_code = 200
+            params = {'channels': {
+                'latest/stable': {'confinement': 'strict'},
+            }}
+
+        self.send_response(status_code)
+        self.send_header('Content-Type', 'text/application+json')
+        self.end_headers()
+        response = json.dumps({'result': [params]}).encode()
+        self.wfile.write(response)
+
+
+class UnixHTTPServer(socketserver.UnixStreamServer):
+
+    def get_request(self):
+        request, client_address = self.socket.accept()
+        # BaseHTTPRequestHandler expects a tuple with the client address at
+        # index 0, so we fake one
+        if len(client_address) == 0:
+            client_address = (self.server_address,)
+        return (request, client_address)
+
+
+class FakeSnapd(fixtures.Fixture):
+
+    def _setUp(self):
+        snapd_fake_socket_path = tempfile.mkstemp()[1]
+        os.unlink(snapd_fake_socket_path)
+
+        socket_path_patcher = mock.patch(
+            'snapcraft.internal.repo.snaps.get_snapd_socket_path_template')
+        mock_socket_path = socket_path_patcher.start()
+        mock_socket_path.return_value = 'http+unix://{}/v2/{{}}'.format(
+            snapd_fake_socket_path.replace('/', '%2F'))
+        self.addCleanup(socket_path_patcher.stop)
+
+        self._start_fake_server(snapd_fake_socket_path)
+
+    def _start_fake_server(self, socket):
+        self.server = UnixHTTPServer(socket, FakeSnapdServer)
+        server_thread = Thread(target=self.server.serve_forever)
+        server_thread.start()
+        self.addCleanup(self._stop_fake_server, server_thread)
+
+    def _stop_fake_server(self, thread):
+        self.server.shutdown()
+        self.server.socket.close()
+        thread.join()
+
+
+class SnapPackageBaseTestCase(tests.TestCase):
+
+    def setUp(self):
+        super().setUp()
+
+        self.useFixture(FakeSnapd())
+
+
+class SnapPackageCurrentChannelTest(SnapPackageBaseTestCase):
+
+    scenarios = [
+        ('stable',
+         dict(snap='fake-snap-stable/stable',
+              expected='latest/stable')),
+        ('latest/stable',
+         dict(snap='fake-snap-stable/latest/stable',
+              expected='latest/stable')),
+        ('candidate/branch',
+         dict(snap='fake-snap-branch/candidate/branch',
+              expected='latest/candidate/branch')),
+        ('track/stable/branch',
+         dict(snap='fake-snap-track-stable-branch/trakc/stable/branch',
```

Comment:
There is a typo here: trakc/track.
This is one of the good reasons to migrate out of those hard-coded fake servers.

Reply:
Fixed, but in the case of this test fixture the request for the _future_ is never made, if I had called it, I would of gotten an error (like `install` or `is_classic`)